internal/kv/csv_file.go

```diff
@@ -0,0 +1,192 @@
+	// in rows in the CSV file.
+	//
+	// This is optional and defaults to comma (",").
+	Delimiter string `json:"delimiter"`
+	// Header overrides the header in the CSV file.
+	//
+	// This is optional and defaults to using the first line of the CSV file as the
+	// header.
+	Header string `json:"header"`
+	mu     sync.Mutex
+	items  map[string]map[string]interface{}
+}
+
+func (store *kvCSVFile) String() string {
+	return toString(store)
+}
+
+// Get retrieves a value from the store.
+func (store *kvCSVFile) Get(ctx context.Context, key string) (interface{}, error) {
+	store.mu.Lock()
+	defer store.mu.Unlock()
+
+	if val, ok := store.items[key]; ok {
+		return val, nil
+	}
+
+	return nil, nil
+}
+
+// Set is unused because this is a read-only store.
+func (store *kvCSVFile) Set(ctx context.Context, key string, val interface{}) error {
+	return errSetNotSupported
+}
+
+// SetWithTTL is unused because this is a read-only store.
+func (store *kvCSVFile) SetWithTTL(ctx context.Context, key string, val interface{}, ttl int64) error {
+	return errSetNotSupported
+}
+
+// IsEnabled returns true if the store is ready for use.
+func (store *kvCSVFile) IsEnabled() bool {
+	return store.items != nil
+}
+
+// Setup creates the store by reading the CSV file into memory.
+func (store *kvCSVFile) Setup(ctx context.Context) error {
+	store.mu.Lock()
+	defer store.mu.Unlock()
+
+	// avoids unnecessary setup
+	if store.items != nil {
+		return nil
+	}
+
+	store.items = make(map[string]map[string]interface{})
+
+	path, err := file.Get(ctx, store.File)
+	defer os.Remove(path)
+	if err != nil {
+		return fmt.Errorf("kv: csv_file: %v", err)
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("kv: csv_file: %v", err)
+	}
+
+	defer f.Close()
+
+	var reader *csv.Reader
+	// the first line of the CSV file is replaced with header if it exists
+	if store.Header != "" {
+		buf, err := bufio.NewReader(f).ReadString('\n')
+		if err != nil {
+			return fmt.Errorf("kv: csv_file: %v", err)
+		}
+		if _, err = f.Seek(int64(len(buf)), io.SeekStart); err != nil {
+			return fmt.Errorf("kv: csv_file: %v", err)
+		}
+
+		h := strings.NewReader(fmt.Sprintf("%s\n", store.Header))
+		reader = csv.NewReader(io.MultiReader(h, f))
+	} else {
+		reader = csv.NewReader(f)
+	}
+
+	if store.Delimiter == "" {
+		store.Delimiter = ","
+	}
+
+	// CSV reader only accepts runes for the comma / delimiter
+	r, _ := utf8.DecodeRune([]byte(store.Delimiter))
+	reader.Comma = r
+
+	// any errors in the CSV file are raised here
+	rows, err := reader.ReadAll()
+	if err != nil {
+		return fmt.Errorf("kv: csv_file: %v", err)
+	}
+
+	var header []string
+	for i, row := range rows {
+		if i == 0 {
+			for i := 0; i < len(row); i++ {
+				header = append(header, row[i])
+			}
+		} else {
+			// the KV store key is the column's value
+			var key string
+			for i := 0; i < len(row); i++ {
+				if header[i] != store.Column {
+					continue
+				}
+
+				key = row[i]
+			}
```

Comment:
Should we return an error if the target column isn't found during setup? If this happens all lookups after are going to be silently empty.

Reply:
done!